app/src/androidTest/java/com/epfl/drawyourpath/database/FirebaseDatabaseTest.kt

```diff
@@ -0,0 +1,211 @@
+
+            override fun getResult(): T = value!!
+
+            override fun <X : Throwable?> getResult(p0: Class<X>): T = value!!
+
+            override fun isCanceled(): Boolean = false
+
+            override fun isComplete(): Boolean = true
+
+            override fun isSuccessful(): Boolean = exception == null
+
+            override fun addOnSuccessListener(p0: Executor, p1: OnSuccessListener<in T>): Task<T> = addOnSuccessListener(p1)
+
+            override fun addOnSuccessListener(p0: Activity, p1: OnSuccessListener<in T>): Task<T> = addOnSuccessListener(p1)
+        }
+    }
+
+    private fun mockDatabaseWithUser(userData: UserData, databaseException: Exception? = null): DatabaseReference {
+        val database = mock(DatabaseReference::class.java)
+        val usersRoot = mock(DatabaseReference::class.java)
+        val userRoot = mock(DatabaseReference::class.java)
+
+        val userProfile = mock(DatabaseReference::class.java)
+
+        // Mock username entry
+        val username = mock(DatabaseReference::class.java)
+        val usernameSnapshot = mockSnapshot(userData.username)
+
+        // TODO: mock more props for other tests
+
+        `when`(username.get()).thenReturn(mockTask(usernameSnapshot, databaseException))
+        `when`(userProfile.child(FirebaseKeys.USERNAME)).thenReturn(username)
+        `when`(userRoot.child(FirebaseKeys.PROFILE)).thenReturn(userProfile)
+        `when`(userRoot.updateChildren(any())).thenReturn(mockTask(null, databaseException))
+        `when`(usersRoot.child(userData.userId!!)).thenReturn(userRoot)
+        `when`(database.child(FirebaseKeys.USERS_ROOT)).thenReturn(usersRoot)
+
+        return database
+    }
+
+    // We can use Mockito to directly mock the Firebase Database without the mock.
+
+    @Test
+    fun isUserInDatabaseForExistingUserReturnsTrue() {
+        val user = UserData(
+            userId = "uid",
+            username = "uname",
+        )
+        val databaseRef = mockDatabaseWithUser(user)
+
+        val res = FirebaseDatabase(databaseRef).isUserInDatabase(user.userId!!).get()
+
+        assertThat(res, `is`(true))
+    }
+
+    @Test
+    fun isUserInDatabaseForNonExistingUserReturnsFalse() {
+        val user = UserData(
+            userId = "uid",
+        )
+        val databaseRef = mockDatabaseWithUser(user)
+
+        val res = FirebaseDatabase(databaseRef).isUserInDatabase(user.userId!!).get()
+
+        assertThat(res, `is`(false))
+    }
+
+    @Test
+    fun isUserInDatabaseWithFailingDatabaseThrows() {
+        val user = UserData(
+            userId = "uid",
+            username = "uname",
+        )
+        val databaseRef = mockDatabaseWithUser(user, Exception("Foobar"))
+
+        assertThrows(Throwable::class.java) {
+            FirebaseDatabase(databaseRef).isUserInDatabase(user.userId!!).get()
+        }
+    }
+
+    @Test
+    fun getUsernameReturnsUsername() {
+        val user = UserData(
+            userId = "uid",
```

Comment:
Maybe you can extract the user outside this function since it used in lots of other test.
Create a variable : var MOCK_USER: UserData = UserData(....)

Reply:
I would say it might be better to keep the data specific to the test to prevent side effects when adding more tests in the database. It might be worth it to do so for bigger tests though. Do you agree?

Reply:
Yes, I'm agree with you in this case because a complete UserData will contains useless data for example for this test (it will contains first name useless for this test  for example):